### ISSUE TITLE
Objective function for flows without variable_costs

### ISSUE THREAD

**Issue by uvchik** [oemof/oemof#337]
If a flow does not have the `variable_costs` defined it does not make sense to add it to the objective function. Nevertheless we loop through every time step.

```python
for i, o in m.FLOWS:
        for t in m.TIMESTEPS:
            # add variable costs
            if m.flows[i, o].variable_costs[0] is not None:
                variable_costs += (m.flow[i, o, t] * m.timeincrement[t] *
                                   m.flows[i, o].variable_costs[t])
```

Wouldn't it be consistent to create a set with all flows with costs set as we do it with the constraint definition?

However, we should check for `variable_costs` before the time-loop.

@oemof/oemof-solph Do you agree, or did I forget something?

**Comment by ckaldemeyer** [comment]
Another option would be to loop only if variable costs are set:

    for i, o in m.FLOWS if m.flows[i, o].variable_costs[0] is not None:
        for t in m.TIMESTEPS:
             # bla

We could also use `hasattr()` to check this.

**Comment by uvchik** [comment]
> Another option would be to loop only if variable costs are set:

That's what I meant with "However, we should check for variable_costs before the time-loop.".

But it is still inconsistent because for all constraints we build sets and loop over the sets, so maybe we should keep this way for all loops.

**Comment by ckaldemeyer** [comment]
> Another option would be to loop only if variable costs are set:
> That's what I meant with "However, we should check for variable_costs before the time-loop.".
> 
> But it is still inconsistent because for all constraints we build sets and loop over the sets, so maybe we should keep this way for all loops.

I agree but wouldn't spend time on this task speaking for myself.

**Comment by simnh** [comment]
I guess it's enough for now to not create a new set, I will move the if condition a few lines up.

**Comment by uvchik** [comment]
Of course it is enough but it is still confusing.

**Comment by uvchik** [comment]
Is there a reason why we sometimes create a set and sometimes not?

**Comment by miraStud** [comment]
Hello everyone I am new here and unfortunately it is not possible to contact you with a private message. My issue is a little of the topic, but has to do with the objective function. 
I want to define a quality factor (exergetic efficiency) to all nodes (busses and components), with the aim to minimize the difference of the input flow and the output flow (exergy loss) of the whole energy system instead of minimizing the costs.
As a first step my system is just comprised of sources, sinks and busses (no transformers) and now my problem is to access the inflowing and outflowing flows to a bus. 
I tried to add following objective_expression to the block Bus:

` def _objective_expression(self, group=None):

        m = self.parent_block()
        exergy_loss = 0 
        if group is None:
            return None

        for n in group:
            for i, o in m.FLOWS:
                for t in m.TIMESTEPS:
                    if m.flows[i, o].exergy_factor[0] is not None: 
                            exergy_loss += (m.flow[i, n, t] * m.timeincrement[t] *
                                             m.flows[i, o].exergy_factor[t])
                            exergy_loss += (- m.flow[n, o, t] * m.timeincrement[t] *
                                            m.flows[i, o].exergy_factor[t])

        self.exergy_loss = Expression(expr=exergy_loss)
        return exergy_loss`

Any suggestions?

**Comment by uvchik** [comment]
Actually it is possible to contact us if you visit our website. But for questions like this we have a forum and I would therefore ask you to copy your question there as it is really does not belong to this issue.

https://forum.openmod-initiative.org/tags/c/qa/oemof

We will try to help you there 😄 

By the way cost are not necessary monetary costs but can also be emissions or even exergy factors.

**Comment by miraStud** [comment]
Ok, thank you!
I opened a new topic:
https://forum.openmod-initiative.org/t/optimizing-the-loss-of-exergy/472

**Comment by uvchik** [comment]
It is not urgent, but I would like to understand which way is better and why we use sets in some cases and if statements in other cases. Therefore I will postpone it.

**Comment by jnnr** [comment]
If you are ok, I will postpone again. Or rather close?

**Comment by simnh** [comment]
Relates to internal refactoring...nothing important for now, as functionality works fine. Postpone is the right way I think.

**Comment by uvchik** [comment]
I will try to check it on Friday otherwise I will postpone it to v0.3.1

**Comment by p-snft** [comment]
If this is not important enough to be tackled, I would suggest to add a "won't fix" tag and close the issue.

**Comment by p-snft** [comment]
As this seems not to be important enough to be tackled, I suggest to be so honest to add the wontfix tag and close this issue.
